### ISSUE TITLE
Automate releases

### ISSUE THREAD

**Issue by mattwynne** [cucumber/build#42]
This effort is almost there. We have a PR being created (e.g. #43) when you update the `CHANGELOG.md` file, adding a new numbered version header.

There are a couple of snags, however, to do with the current solution of using a `VERSION` file to represent the version to be released / last released.

- In `pre-release.yaml`, read the current release version from the git tags instead. This will mean we don't need to make a new commit in the release PR, and so we don't need to worry about GPG signing that commit. In fact we can rebase rather than squash since the `main` branch has a rule to enforce linear history as well as signed commits. That will make the release PR nicer to review.
- in `release.yaml`, just read the version to be released from `changelog latest`.

TODO:

- [x] re-enable signed commits check for release branch
- [x] update the docs
- [x] demo / review the process with the core team 
- [x] deal with [the issue](https://github.com/cucumber/build/issues/42#issuecomment-888565950) where pushes to `main` occur _during_ a release.
- [x] deal with the problem rebasing squashed commits

**Comment by mattwynne** [comment]
Okay, the work on removing the `VERSION` file is done, and appears to have worked.

The rebasing idea has not, however. Old commits that have been squashed into the release branch are being rebased on, from before we had the commit signing rule on main.

Perhaps we need to re-branch the `release` branch from main as of the last release.

**Comment by mattwynne** [comment]
alright well I tried deleting the `release` branch and creating it again from the v0.5.2 tag but the new PR still has a bunch of commits I don't recognise. Need to have another look at this when I have more time.

**Comment by aurelien-reeves** [comment]
I would like to enforce the security of automated releases.

To do so, I would like to set action permissions, in the "Actions" settings, to "local actions only", or maybe "Allow select actions" with only the ones from github.

![image](https://user-images.githubusercontent.com/24386780/127290430-d68b78cc-7db3-4b79-baa3-e87aaaf8ed76.png)

External actions are a [supply chain attack](https://www.upguard.com/blog/supply-chain-attack) vector. That would result in a breach into the cucumber build mechanism which is used to release all cucumber things.

Preventing this would mean, in `pre-release.yml`, avoid reading the changelog with an external action, and avoid creating the realease PR with another one.

`mattwynne/changelog-action` is a fork, based on the docker image `mattwynne/changelog` which is also a fork with a binary in it written in go. A lot can go wrong here. 

And `peter-evans/create-pull-request` is totally out of sight.

In `release.yml`, there are also a lot of dependencies out of our sight. At the moment as the release still require some human review, that is barely an issue. But as it is planned to automate all the things, I think we should be more careful with those dependencies.

**Comment by mattwynne** [comment]
Aurélin and I paired today. We did some mitigation of the chances of a supply chain attack on the GitHub Actions plugins by choosing the "Allow select actions" and adding the two 3rd party ones we're using to the list.

Ultimately, we think it might be best to fork those 3rd party plugins to the Cucumber org, and choose the "Use local actions only" option.

**Comment by mattwynne** [comment]
We had to do some tidying up of the `release` branch, as the previous mechanism had made a bit of a mess of it. We ended up re-creating the branch from the e2dc1111de3ea967fab9bcb5b57c6dc00a4b9a65 commit, which is where we think the v0.5.2 tag _should_ have been pointing.

**Comment by mattwynne** [comment]
With the `release` branch tidied up, the pre-release workflow successfully created https://github.com/cucumber/build/pull/44. We reviewed and merged the PR, so the first "automated" release is now in progress!

**Comment by mattwynne** [comment]
We have an issue that, if the pre-release script runs _while a release is still in progress_ (i.e. you push a new commit to `main` before the new release's `vx.y.z` tag has been pushed) then it will still trigger, thinking that the release has not happened yet:

e.g. https://github.com/cucumber/build/runs/3185235101

I guess we need a way to signal to the script that the release is in progress, too. Maybe we can add the git tag at the beginning of the `release.yaml` instead of at the end?

**Comment by mattwynne** [comment]
Actually, while I think that is a real issue, the cause of this build failure seems to be something else, to do with a merge conflict when rebasing `main` onto the `release` branch. It seems like we still haven't got this all square.

I can't help thinking that having one branch per release (e.g. `release/0.6.0`), rather than having a single long-lived `release` branch, would help.

**Comment by mattwynne** [comment]
Trying out one branch per release in 8701c1c7854bdd6f5db62325d2c23e0b646a501b

**Comment by mattwynne** [comment]
Same problem, since I think the root cause is the squash commit going onto the release branch, which confuses git.

Maybe we need a commit that merges back from the release branch to main?

**Comment by aurelien-reeves** [comment]
I may have found a working and simpler workflow based on a single `release` branch and using only merge. No rebase, no squash.

TL;DR:
- checkout a brand new `release` branch from https://github.com/cucumber/build/commit/7854135920aa38b55acc4934c8c67db001883dac
- when pre-releasing, checkout a new `pre-release` branch from main
- create a PR for the `pre-release` based on the `release` branch
- to release, merge the `pre-release` branch on `release`, and tag the current commit

Long version:

We would have to recreate a release branch from the following commit, because this is the last one on master before 0.6.0: https://github.com/cucumber/build/commit/7854135920aa38b55acc4934c8c67db001883dac

(the tag v0.6.0 is not usable: it point to a commit that does not exists on main).

So that would looks like this:

```shell
$ git show-branch
! [main] Update README.md
 * [release] Tweak working in release PR
--
+  [main] Update README.md
+  [main^] Update docs for automated release workflow (#47)
+  [main~2] Introduce Docker layer cache (#46)
+  [main~3] Report version numbers detected
+  [main~4] Fix
+  [main~5] Fix copy/paste error
+  [main~6] Parse version number correctly from changelog
+  [main~7] YAML
+  [main~8] Not sure why this needs an author
+  [main~9] Always take changes from `main` if there's a conflict
+  [main~10] One-branch-per-release was not finished in the previous commit
+  [main~11] Use one branch per release
+  [main~12] Trigger build
+  [main~13] Fix CHANGELOG titles
+  [main~14] Use more consistent name
+  [main~15] Update CHANGELOG for #35
+  [main~16] Merge #35
+  [main~17] Add CHANGELOG entry for https://github.com/cucumber/build/pull/38
+  [main~18] Update Perl dependencies (#38)
+  [main~19] Fix links in CHANGELOG
+  [main~20] Fix diff link in 6.0.0 header
+* [release] Tweak working in release PR
```

To pre-release, simply checkout a new `pre-release` branch from `main`

```shell
$ git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

$ git checkout -b pre-release
Switched to a new branch 'pre-release'

$ git show-branch
! [main] Update README.md
 * [pre-release] Update README.md
  ! [release] Tweak working in release PR
---
+*  [main] Update README.md
+*  [main^] Update docs for automated release workflow (#47)
+*  [main~2] Introduce Docker layer cache (#46)
+*  [main~3] Report version numbers detected
+*  [main~4] Fix
+*  [main~5] Fix copy/paste error
+*  [main~6] Parse version number correctly from changelog
+*  [main~7] YAML
+*  [main~8] Not sure why this needs an author
+*  [main~9] Always take changes from `main` if there's a conflict
+*  [main~10] One-branch-per-release was not finished in the previous commit
+*  [main~11] Use one branch per release
+*  [main~12] Trigger build
+*  [main~13] Fix CHANGELOG titles
+*  [main~14] Use more consistent name
+*  [main~15] Update CHANGELOG for #35
+*  [main~16] Merge #35
+*  [main~17] Add CHANGELOG entry for https://github.com/cucumber/build/pull/38
+*  [main~18] Update Perl dependencies (#38)
+*  [main~19] Fix links in CHANGELOG
+*  [main~20] Fix diff link in 6.0.0 header
+*+ [release] Tweak working in release PR
```
`main` and `pre-release` have the same HEAD
`release` still has the HEAD from the previous release

Now, if we create a PR to merge `pre-release` on `release`, we will be able to review the diff between the previous release and the new proposed one:

```shell
$ git log --oneline release..pre-release
b2f594b (origin/main, origin/HEAD, pre-release, main) Update README.md
e9fc872 Update docs for automated release workflow (#47)
fec71a8 Introduce Docker layer cache (#46)
2a821bd Report version numbers detected
486f167 Fix
89cea8a Fix copy/paste error
f91c4a8 Parse version number correctly from changelog
a2b919f YAML
d61fd21 Not sure why this needs an author
e96d523 Always take changes from `main` if there's a conflict
99af31b One-branch-per-release was not finished in the previous commit
8701c1c Use one branch per release
2233ee6 Trigger build
177550d Fix CHANGELOG titles
ff8aaee Use more consistent name
bdc3157 Update CHANGELOG for #35
0e8c7e7 Merge #35
c61991f Add CHANGELOG entry for https://github.com/cucumber/build/pull/38
bf25f4d Update Perl dependencies (#38)
f4cefd3 Fix links in CHANGELOG
898cb77 Fix diff link in 6.0.0 header
```

To finalize the release, just merge pre-release into release. No rebase (so no squash), just merge.

```shell
$ git merge pre-release
Updating 7854135..b2f594b
Fast-forward
 .github/workflows/build.yaml                        | 22 +++++++++++++++++++++-
 .github/workflows/pre-release.yaml                  | 25 +++++++++++++++++++------
 .github/workflows/release.yaml                      |  2 +-
 CHANGELOG.md                                        | 35 +++++++++++++++++++----------------
 Dockerfile                                          |  9 ++++-----
 Makefile                                            |  6 +++++-
 README.md                                           | 41 +++++++----------------------------------
 scripts/{download-chrome.sh => install-chromium.sh} |  2 +-
 8 files changed, 77 insertions(+), 65 deletions(-)
 rename scripts/{download-chrome.sh => install-chromium.sh} (95%)

$ git show-branch
! [main] Update README.md
 ! [pre-release] Update README.md
  * [release] Update README.md
---
++* [main] Update README.md
```

`release` now points to the last commit we wanted to actually release. `main` and `release` have the same history, and we can add a tag to that released commit. The HEAD of release is the same commit as the one that has been released from main, not another one. That way we can reproduce the process for the next time.

```shell
$ git checkout main
Already on 'main'
Your branch is up to date with 'origin/main'.

$ git show --compact-summary
commit b2f594bc26d15604460b791de11356a1f3429bbe (HEAD -> main, origin/main, origin/HEAD, release, pre-release)
# ...
```

**Comment by mattwynne** [comment]
I think we should abandon the 3rd party pull request action and use `gh`. It's natively supported in GitHub Actions anyway. I've done a proof of concept here: https://github.com/mattwynne/release-workflow-test

I think it will work much better, but we still have the nagging problem of the potential that someone could use a squash merge commit to the release branch and fuck it all up.

Maybe we should try a different approach, and look into using _deployments_ instead. https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments

**Comment by mattwynne** [comment]
I really like the PR-based approval step in the workflow. The pull request allows us to clearly see exactly what's going into the release.

However, this model creates some tension:

1) We'd like to enforce linear, signed commits on `main`, by allowing feature branches to be squashed by the core team member who merges the change, instead of having them make a merge commit.
2) GitHub's rules about merge strategies are per-repo, not per-branch, so this also means release PRs would end up as squashed commits to the `release` branch too. There's no way I can see to prevent this, other than by convention.
3) If we end up with a squashed commit to the `release` branch, we break the git history on that branch. This makes subsequent releases hard, as git will try to merge the commits from previous releases into that branch again, potentially creating merge conflicts, but also making the release PR hard to read as it contains more commits than are really being released. Also, if a tag is put on the released commit, it's not showing on `main` where the original commit was.

Ideally, I think, we'd have squash merges to `main` and merge commits to `release`, but GitHub won't let us enforce this.

**Comment by mattwynne** [comment]
I tried out the option of release approval but it's pretty lame. You have no way to see any details of what it is you're approving.

**Comment by mattwynne** [comment]
OK I have something that seems to work in https://github.com/mattwynne/release-workflow

It works by creating two branches in the `pre-release` workflow:

* a `pre-release/vx.y.z` branch off `main` when you trigger the release
* a `release/vx.y.z` branch off the _previous_ release version

Then it makes a PR onto that `release/vx.y.z` branch from the `pre-release/vx.y.z` branch that contains the new commits.

Once the PR is merged into the `release/*` branch, the `release` workflow runs, creates a GitHub Release, and even if there's a squash merge commit, that still seems to tag the commit on `main`, rather than the throwaway squash commit.

I think the next step might be to factor out some re-usable (and perhaps even testable!) actions from that spike, so that we can use them here.

**Comment by mattwynne** [comment]
I've updated the `pre-release` and `release` actions workflows in this repo, and the `pre-release` one [seems to have worked](https://github.com/cucumber/build/pull/52).

I had some trouble bootstrapping it:

1. I had to move the tag 😲 for v0.6.0 as it was pointing to one of those ugly squashed commits, so I moved it back to the [relevant commit](https://github.com/cucumber/build/commit/b34ac5abc308f4e3247f5a040e80ff5a3afa8f8c) on `main`.
2. creating the `release/0.7.0` branch failed with `refusing to allow a GitHub App to create or update workflow ".github/workflows/pre-release.yaml" without "workflows" permission` which I *think* is because the diff for this release includes changes to the workflow file, which the bot is not allowed to make. This could trip us up in future, assuming we need to revise the workflows again. I worked around it by manually pushing the `release/0.7.0` branch from the (corrected) `v0.6.0` tag.

**Comment by mattwynne** [comment]
I've merged the PR for the release, so now we can see if the `release` workflow works...

**Comment by mattwynne** [comment]
OK the release workflow has worked!

I'd like to leave this open until we make another release, just in case.

**Comment by mattwynne** [comment]
It's working.
